### PR TITLE
feat(api/gateway): Implement context cancellation middleware

### LINKED ISSUE #726: fix(gateway): Ensure r.Context is getting canceled once the processing of the request ends with error or success or anything else.
Pls double check that this Context is getting canceled once the processing of the request ends with error or success or anything else. If is not closed automatically, then we need to add to each handler:
```go
ctx, cancel := context.WithCancel(r.Context())
defer cancel()
```

_Originally posted by @Wondertan in https://github.com/celestiaorg/celestia-node/pull/723#discussion_r874813310_

### PR DESCRIPTION
**Pull request by renaynay**
Resolves #726

### CODE REVIEW

File: api/gateway/middleware.go [hunk 1]
```diff
@@ -34,3 +37,13 @@ func checkPostDisabled(state state.Module) mux.MiddlewareFunc {
 		})
 	}
 }
+
+// wrapRequestContext ensures we implement a deadline on serving requests
+// via the gateway server-side to prevent context leaks.
+func wrapRequestContext(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx, cancel := context.WithTimeout(r.Context(), time.Minute)
```
Comment (Wondertan) [inline api/gateway/middleware.go:45]: N: extract the timeout in a separate func

**Review by Wondertan — COMMENTED** [review]
LG

**Review by distractedm1nd — DISMISSED** [review]

**Review by distractedm1nd — DISMISSED** [review]
already reviewed, reapproving

**Review by distractedm1nd — APPROVED** [review]

File: api/gateway/middleware.go (hunk 1 above)
Comment (renaynay) [reply api/gateway/middleware.go]: ok pls approve now :(

**Review by Wondertan — APPROVED** [review]

**Pull request merged**